### PR TITLE
avoid WOperator with StaticArrays

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Since WOperator is mutable and allocates.

```julia
using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function rober(u,p,t)
  y₁,y₂,y₃ = u
  k₁,k₂,k₃ = p
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
  SA[dy₁,dy₂,dy₃]
end
prob = ODEProblem{false}(rober,SA[1.0,0.0,0.0],(0.0,1e5),SA[0.04,3e7,1e4])
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{3}()),save_everystep=false) 
# Before:
# 14.800 μs (26 allocations: 3.28 KiB)
# After:
# 8.900 μs (26 allocations: 3.28 KiB)

function rober_jac(u,p,t)
  y₁,y₂,y₃ = u
  k₁,k₂,k₃ = p
  J11 = k₁ * -1
  J21 = k₁
  J31 = 0
  J12 = y₃ * k₃
  J22 = y₂ * k₂ * -2 + y₃ * k₃ * -1
  J32 = y₂ * 2 * k₂
  J13 = k₃ * y₂
  J23 = k₃ * y₂ * -1
  J33 = 0
  SA[J11 J12 J13
     J21 J22 J23
     J31 J32 J33]
end

f = ODEFunction(rober, jac=rober_jac)
prob_jac = ODEProblem(f,SA[1.0,0.0,0.0],(0.0,1e5),SA[0.04,3e7,1e4])
@btime sol = solve(prob_jac,Rosenbrock23(chunk_size = Val{3}()),save_everystep=false) # 9.400 μs (87 allocations: 16.47 KiB)

# Before:
# 9.400 μs (87 allocations: 16.47 KiB)
# After:
# 8.000 μs (27 allocations: 2.98 KiB)
```

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
```julia
using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function rober(u,p,t)
  y₁,y₂,y₃ = u
  k₁,k₂,k₃ = p
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
  SA[dy₁,dy₂,dy₃]
end
prob = ODEProblem{false}(rober,SA[1.0,0.0,0.0],(0.0,1e5),SA[0.04,3e7,1e4])
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,TRBDF2(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,KenCarp4(chunk_size = Val{3}()),save_everystep=false)
# Before:
# 91.500 μs (124 allocations: 7.22 KiB)
# 85.100 μs (246 allocations: 10.95 KiB)

# After:
# 59.200 μs (124 allocations: 6.94 KiB)
# 59.500 μs (246 allocations: 10.67 KiB)
```

**Comment by ChrisRackauckas** [comment]
```julia
using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires(u,p,t)
  y1,y2,y3,y4,y5,y6,y7,y8 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y6 + 0.43*y7
  dy6 = -280.0*y6*y8 + 0.69*y4 + 1.71*y5 -
           0.43*y6 + 0.69*y7
  dy7 = 280.0*y6*y8 - 1.81*y7
  dy8 = -280.0*y6*y8 + 1.81*y7
  SA[dy1,dy2,dy3,dy4,dy5,dy6,dy7,dy8]
end

u0 = SA[1,0,0,0,0,0,0,0.0057]
prob = ODEProblem(hires,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,TRBDF2(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,KenCarp4(chunk_size = Val{3}()),save_everystep=false)
# Before:
# 88.400 μs (26 allocations: 8.14 KiB)
# 113.500 μs (71 allocations: 7.73 KiB)
# 191.500 μs (191 allocations: 11.34 KiB)

# After:
# 127.000 μs (26 allocations: 6.36 KiB)
# 173.500 μs (71 allocations: 7.03 KiB)
# 291.300 μs (191 allocations: 10.64 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_5(u,p,t)
  y1,y2,y3,y4,y5 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y2 + 0.43*y4
  SA[dy1,dy2,dy3,dy4,dy5]
end

u0 = SA[1,0,0,0,0.0057]
prob = ODEProblem(hires_5,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{5}()),save_everystep=false)
@btime sol = solve(prob,TRBDF2(chunk_size = Val{5}()),save_everystep=false)
@btime sol = solve(prob,KenCarp4(chunk_size = Val{5}()),save_everystep=false)

# Before:
# 26.900 μs (26 allocations: 4.69 KiB)
# 38.800 μs (67 allocations: 5.64 KiB)
# 60.600 μs (187 allocations: 9.25 KiB)

# After:
# 30.500 μs (26 allocations: 4.03 KiB)
# 40.200 μs (67 allocations: 5.41 KiB)
# 61.600 μs (187 allocations: 9.02 KiB)


using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_4(u,p,t)
  y1,y2,y3,y4 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y2
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  SA[dy1,dy2,dy3,dy4]
end

u0 = SA[1,0,0,0.0057]
prob = ODEProblem(hires_4,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{4}()),save_everystep=false)
@btime sol = solve(prob,TRBDF2(chunk_size = Val{4}()),save_everystep=false)
@btime sol = solve(prob,KenCarp4(chunk_size = Val{4}()),save_everystep=false)

# Before:
# 305.300 μs (6866 allocations: 316.03 KiB)
# 34.000 μs (67 allocations: 5.16 KiB)
# 50.500 μs (187 allocations: 8.77 KiB)

# After:
# 22.400 μs (26 allocations: 3.36 KiB)
# 34.300 μs (70 allocations: 5.14 KiB)
# 50.100 μs (190 allocations: 8.75 KiB)
```

Same cutoff as https://github.com/SciML/OrdinaryDiffEq.jl/pull/1443 . `\` seems fastest at 4 and lower, and `lu` before that.

**Comment by ChrisRackauckas** [comment]
Wow, caching the inv makes `Rodas4` and `Rodas5` amazing.

```julia
using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function rober(u,p,t)
  y₁,y₂,y₃ = u
  k₁,k₂,k₃ = p
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
  SA[dy₁,dy₂,dy₃]
end
prob = ODEProblem{false}(rober,SA[1.0,0.0,0.0],(0.0,1e5),SA[0.04,3e7,1e4])
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,Rodas4(chunk_size = Val{3}()),save_everystep=false)

# Before:
# 14.900 μs (26 allocations: 3.28 KiB)
# 25.800 μs (26 allocations: 4.22 KiB)
# 90.900 μs (124 allocations: 7.22 KiB)
# 83.100 μs (246 allocations: 10.95 KiB)

# After:
# 8.833 μs (26 allocations: 3.03 KiB)
# 13.900 μs (26 allocations: 3.97 KiB)
# 58.800 μs (114 allocations: 6.59 KiB)
# 52.400 μs (237 allocations: 10.34 KiB)

# Using inv version
# 9.100 μs (26 allocations: 3.03 KiB)
# 12.800 μs (26 allocations: 3.97 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires(u,p,t)
  y1,y2,y3,y4,y5,y6,y7,y8 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y6 + 0.43*y7
  dy6 = -280.0*y6*y8 + 0.69*y4 + 1.71*y5 -
           0.43*y6 + 0.69*y7
  dy7 = 280.0*y6*y8 - 1.81*y7
  dy8 = -280.0*y6*y8 + 1.81*y7
  SA[dy1,dy2,dy3,dy4,dy5,dy6,dy7,dy8]
end

u0 = SA[1,0,0,0,0,0,0,0.0057]
prob = ODEProblem(hires,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,TRBDF2(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,KenCarp4(chunk_size = Val{3}()),save_everystep=false)

# Before:
# 88.400 μs (26 allocations: 8.14 KiB)
# 113.500 μs (71 allocations: 7.73 KiB)
# 191.500 μs (191 allocations: 11.34 KiB)

# After:
# 127.000 μs (26 allocations: 6.36 KiB)
# 173.500 μs (71 allocations: 7.03 KiB)
# 291.300 μs (191 allocations: 10.64 KiB)

# After inv:
# 92.000 μs (26 allocations: 6.36 KiB)
# 51.600 μs (26 allocations: 7.61 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_5(u,p,t)
  y1,y2,y3,y4,y5 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y2 + 0.43*y4
  SA[dy1,dy2,dy3,dy4,dy5]
end

u0 = SA[1,0,0,0,0.0057]
prob = ODEProblem(hires_5,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{5}()),save_everystep=false)
@btime sol = solve(prob,Rodas4(chunk_size = Val{5}()),save_everystep=false)
@btime sol = solve(prob,TRBDF2(chunk_size = Val{5}()),save_everystep=false)
@btime sol = solve(prob,KenCarp4(chunk_size = Val{5}()),save_everystep=false)

# Before:
# 27.800 μs (26 allocations: 4.69 KiB)
# 26.100 μs (26 allocations: 5.97 KiB)
# 38.400 μs (67 allocations: 5.64 KiB)
# 60.600 μs (187 allocations: 9.25 KiB)

# After:
# 30.200 μs (26 allocations: 4.03 KiB)
# 36.300 μs (26 allocations: 5.20 KiB)
# 40.200 μs (67 allocations: 5.41 KiB)
# 57.600 μs (187 allocations: 9.02 KiB)

# After inv:
# 23.300 μs (26 allocations: 4.03 KiB)
# 18.900 μs (26 allocations: 5.00 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_4(u,p,t)
  y1,y2,y3,y4 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y2
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  SA[dy1,dy2,dy3,dy4]
end

u0 = SA[1,0,0,0.0057]
prob = ODEProblem(hires_4,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{4}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{4}()),save_everystep=false)

# Before:
#22.200 μs (26 allocations: 3.84 KiB)
#21.600 μs (26 allocations: 5.11 KiB)
#33.600 μs (70 allocations: 5.30 KiB)
#50.400 μs (190 allocations: 8.91 KiB)

# After:
#22.300 μs (26 allocations: 3.36 KiB)
#25.600 μs (26 allocations: 4.59 KiB)
#34.300 μs (70 allocations: 5.14 KiB)
#46.000 μs (190 allocations: 8.75 KiB)

# After inv:
# 11.300 μs (26 allocations: 3.36 KiB)
# 9.400 μs (26 allocations: 4.59 KiB)
```

**Comment by ChrisRackauckas** [comment]
More clear:

```julia
using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function rober(u,p,t)
  y₁,y₂,y₃ = u
  k₁,k₂,k₃ = p
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
  SA[dy₁,dy₂,dy₃]
end
prob = ODEProblem{false}(rober,SA[1.0,0.0,0.0],(0.0,1e5),SA[0.04,3e7,1e4])
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,Rodas4(chunk_size = Val{3}()),save_everystep=false)

# calc_W returns lu(W):
15.000 μs (26 allocations: 3.28 KiB)
25.900 μs (26 allocations: 4.22 KiB)

# calc_W returns W:
9.400 μs (26 allocations: 3.03 KiB)
13.900 μs (26 allocations: 3.97 KiB)

# calc_W returns inv(W):
9.000 μs (26 allocations: 3.03 KiB)
12.800 μs (26 allocations: 3.97 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_4(u,p,t)
  y1,y2,y3,y4 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y2
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  SA[dy1,dy2,dy3,dy4]
end

u0 = SA[1,0,0,0.0057]
prob = ODEProblem(hires_4,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{4}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{4}()),save_everystep=false)

# calc_W returns lu(W):
21.700 μs (26 allocations: 3.84 KiB)
21.400 μs (26 allocations: 5.11 KiB)

# calc_W returns W:
22.200 μs (26 allocations: 3.36 KiB)
25.600 μs (26 allocations: 4.59 KiB)

# calc_W returns inv(W):
11.300 μs (26 allocations: 3.36 KiB)
9.600 μs (26 allocations: 4.59 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_5(u,p,t)
  y1,y2,y3,y4,y5 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y2 + 0.43*y4
  SA[dy1,dy2,dy3,dy4,dy5]
end

u0 = SA[1,0,0,0,0.0057]
prob = ODEProblem(hires_5,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{5}()),save_everystep=false)
@btime sol = solve(prob,Rodas4(chunk_size = Val{5}()),save_everystep=false)

# calc_W returns lu(W):
27.100 μs (26 allocations: 4.69 KiB)
27.300 μs (26 allocations: 5.67 KiB)

# calc_W returns W:
30.200 μs (26 allocations: 4.03 KiB)
35.600 μs (26 allocations: 5.00 KiB)

# calc_W returns inv(W):
23.200 μs (26 allocations: 4.03 KiB)
18.800 μs (26 allocations: 5.00 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires(u,p,t)
  y1,y2,y3,y4,y5,y6,y7,y8 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y6 + 0.43*y7
  dy6 = -280.0*y6*y8 + 0.69*y4 + 1.71*y5 -
           0.43*y6 + 0.69*y7
  dy7 = 280.0*y6*y8 - 1.81*y7
  dy8 = -280.0*y6*y8 + 1.81*y7
  SA[dy1,dy2,dy3,dy4,dy5,dy6,dy7,dy8]
end

u0 = SA[1,0,0,0,0,0,0,0.0057]
prob = ODEProblem(hires,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{8}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{8}()),save_everystep=false)

# calc_W returns lu(W):
89.500 μs (26 allocations: 8.14 KiB)
67.200 μs (26 allocations: 9.22 KiB)

# calc_W returns W:
128.000 μs (26 allocations: 6.36 KiB)
144.600 μs (26 allocations: 7.61 KiB)

# calc_W returns inv(W):
91.700 μs (26 allocations: 6.36 KiB)
52.100 μs (26 allocations: 7.61 KiB)









using OrdinaryDiffEq, StaticArrays, BenchmarkTools
const k1=.35e0
const k2=.266e2
const k3=.123e5
const k4=.86e-3
const k5=.82e-3
const k6=.15e5
const k7=.13e-3
const k8=.24e5
const k9=.165e5
const k10=.9e4
const k11=.22e-1
const k12=.12e5
const k13=.188e1
const k14=.163e5
const k15=.48e7
const k16=.35e-3
const k17=.175e-1
const k18=.1e9
const k19=.444e12
const k20=.124e4
const k21=.21e1
const k22=.578e1
const k23=.474e-1
const k24=.178e4
const k25=.312e1

function f(y,p,t)
 r1  = k1 *y[1]
 r2  = k2 *y[2]*y[4]
 r3  = k3 *y[5]*y[2]
 r4  = k4 *y[7]
 r5  = k5 *y[7]
 r6  = k6 *y[7]*y[6]
 r7  = k7 *y[9]
 r8  = k8 *y[9]*y[6]
 r9  = k9 *y[11]*y[2]
 r10 = k10*y[11]*y[1]
 r11 = k11*y[13]
 r12 = k12*y[10]*y[2]
 r13 = k13*y[14]
 r14 = k14*y[1]*y[6]
 r15 = k15*y[3]
 r16 = k16*y[4]
 r17 = k17*y[4]
 r18 = k18*y[16]
 r19 = k19*y[16]
 r20 = k20*y[17]*y[6]
 r21 = k21*y[19]
 r22 = k22*y[19]
 r23 = k23*y[1]*y[4]
 r24 = k24*y[19]*y[1]
 r25 = k25*y[20]

 dy1  = -r1-r10-r14-r23-r24+
          r2+r3+r9+r11+r12+r22+r25
 dy2  = -r2-r3-r9-r12+r1+r21
 dy3  = -r15+r1+r17+r19+r22
 dy4  = -r2-r16-r17-r23+r15
 dy5  = -r3+r4+r4+r6+r7+r13+r20
 dy6  = -r6-r8-r14-r20+r3+r18+r18
 dy7  = -r4-r5-r6+r13
 dy8  = r4+r5+r6+r7
 dy9  = -r7-r8
 dy10 = -r12+r7+r9
 dy11 = -r9-r10+r8+r11
 dy12 = r9
 dy13 = -r11+r10
 dy14 = -r13+r12
 dy15 = r14
 dy16 = -r18-r19+r16
 dy17 = -r20
 dy18 = r20
 dy19 = -r21-r22-r24+r23+r25
 dy20 = -r25+r24
 SA[dy1,dy2,dy3,dy4,dy5,dy6,dy7,dy8,dy9,dy10,dy11,dy12,dy13,dy14,dy15,dy16,dy17,dy18,dy19,dy20]
end

u0 = zeros(20)
u0[2]  = 0.2
u0[4]  = 0.04
u0[7]  = 0.1
u0[8]  = 0.3
u0[9]  = 0.01
u0[17] = 0.007
u0 = SA[u0...]
prob = ODEProblem(f,u0,(0.0,60.0))
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{8}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{8}()),save_everystep=false)

# calc_W returns lu(W):
3.936 ms (672 allocations: 405.56 KiB)
3.092 ms (1678 allocations: 566.19 KiB)

# calc_W returns W:
12.170 ms (482 allocations: 814.94 KiB)
23.892 ms (922 allocations: 1.54 MiB)

# calc_W returns inv(W):
4.082 ms (216 allocations: 664.12 KiB)
2.995 ms (166 allocations: 496.88 KiB)
```

**Comment by ChrisRackauckas** [comment]
The last thing needs https://github.com/JuliaArrays/StaticArrays.jl/pull/977

**Comment by ChrisRackauckas** [comment]
This now gives a "Best of Both Worlds" and speeds up the small case quite a bit:

```julia
using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function rober(u,p,t)
  y₁,y₂,y₃ = u
  k₁,k₂,k₃ = p
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
  SA[dy₁,dy₂,dy₃]
end
prob = ODEProblem{false}(rober,SA[1.0,0.0,0.0],(0.0,1e5),SA[0.04,3e7,1e4])
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{3}()),save_everystep=false)
@btime sol = solve(prob,Rodas4(chunk_size = Val{3}()),save_everystep=false)

# calc_W returns lu(W):
15.000 μs (26 allocations: 3.28 KiB)
25.900 μs (26 allocations: 4.22 KiB)

# calc_W returns W:
9.400 μs (26 allocations: 3.03 KiB)
13.900 μs (26 allocations: 3.97 KiB)

# calc_W returns inv(W):
9.000 μs (26 allocations: 3.03 KiB)
12.800 μs (26 allocations: 3.97 KiB)

# Now
9.000 μs (26 allocations: 3.03 KiB)
12.900 μs (26 allocations: 3.97 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_4(u,p,t)
  y1,y2,y3,y4 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y2
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  SA[dy1,dy2,dy3,dy4]
end

u0 = SA[1,0,0,0.0057]
prob = ODEProblem(hires_4,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{4}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{4}()),save_everystep=false)

# calc_W returns lu(W):
21.700 μs (26 allocations: 3.84 KiB)
21.400 μs (26 allocations: 5.11 KiB)

# calc_W returns W:
22.200 μs (26 allocations: 3.36 KiB)
25.600 μs (26 allocations: 4.59 KiB)

# calc_W returns inv(W):
11.300 μs (26 allocations: 3.36 KiB)
9.600 μs (26 allocations: 4.59 KiB)

# Now
11.200 μs (26 allocations: 3.36 KiB)
9.400 μs (26 allocations: 4.59 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires_5(u,p,t)
  y1,y2,y3,y4,y5 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y2 + 0.43*y4
  SA[dy1,dy2,dy3,dy4,dy5]
end

u0 = SA[1,0,0,0,0.0057]
prob = ODEProblem(hires_5,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{5}()),save_everystep=false)
@btime sol = solve(prob,Rodas4(chunk_size = Val{5}()),save_everystep=false)

# calc_W returns lu(W):
27.100 μs (26 allocations: 4.69 KiB)
27.300 μs (26 allocations: 5.67 KiB)

# calc_W returns W:
30.200 μs (26 allocations: 4.03 KiB)
35.600 μs (26 allocations: 5.00 KiB)

# calc_W returns inv(W):
23.200 μs (26 allocations: 4.03 KiB)
18.800 μs (26 allocations: 5.00 KiB)

# Now
23.000 μs (26 allocations: 4.03 KiB)
18.900 μs (26 allocations: 5.00 KiB)

using OrdinaryDiffEq, StaticArrays, BenchmarkTools
function hires(u,p,t)
  y1,y2,y3,y4,y5,y6,y7,y8 = u
  dy1 = -1.71*y1 + 0.43*y2 + 8.32*y3 + 0.0007
  dy2 = 1.71*y1 - 8.75*y2
  dy3 = -10.03*y3 + 0.43*y4 + 0.035*y5
  dy4 = 8.32*y2 + 1.71*y3 - 1.12*y4
  dy5 = -1.745*y5 + 0.43*y6 + 0.43*y7
  dy6 = -280.0*y6*y8 + 0.69*y4 + 1.71*y5 -
           0.43*y6 + 0.69*y7
  dy7 = 280.0*y6*y8 - 1.81*y7
  dy8 = -280.0*y6*y8 + 1.81*y7
  SA[dy1,dy2,dy3,dy4,dy5,dy6,dy7,dy8]
end

u0 = SA[1,0,0,0,0,0,0,0.0057]
prob = ODEProblem(hires,u0,(0.0,321.8122))
# Defaults to reltol=1e-3, abstol=1e-6
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{8}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{8}()),save_everystep=false)

using Profile
@profile for i in 1:1000 sol = solve(prob,Rosenbrock23(chunk_size = Val{8}()),save_everystep=false) end
Juno.profiler()

# calc_W returns lu(W):
89.500 μs (26 allocations: 8.14 KiB)
67.200 μs (26 allocations: 9.22 KiB)

# calc_W returns W:
128.000 μs (26 allocations: 6.36 KiB)
144.600 μs (26 allocations: 7.61 KiB)

# calc_W returns inv(W):
91.700 μs (26 allocations: 6.36 KiB)
52.100 μs (26 allocations: 7.61 KiB)

# Now
88.800 μs (26 allocations: 8.14 KiB)
66.900 μs (26 allocations: 9.22 KiB)





using OrdinaryDiffEq, StaticArrays, BenchmarkTools
const k1=.35e0
const k2=.266e2
const k3=.123e5
const k4=.86e-3
const k5=.82e-3
const k6=.15e5
const k7=.13e-3
const k8=.24e5
const k9=.165e5
const k10=.9e4
const k11=.22e-1
const k12=.12e5
const k13=.188e1
const k14=.163e5
const k15=.48e7
const k16=.35e-3
const k17=.175e-1
const k18=.1e9
const k19=.444e12
const k20=.124e4
const k21=.21e1
const k22=.578e1
const k23=.474e-1
const k24=.178e4
const k25=.312e1

function f(y,p,t)
 r1  = k1 *y[1]
 r2  = k2 *y[2]*y[4]
 r3  = k3 *y[5]*y[2]
 r4  = k4 *y[7]
 r5  = k5 *y[7]
 r6  = k6 *y[7]*y[6]
 r7  = k7 *y[9]
 r8  = k8 *y[9]*y[6]
 r9  = k9 *y[11]*y[2]
 r10 = k10*y[11]*y[1]
 r11 = k11*y[13]
 r12 = k12*y[10]*y[2]
 r13 = k13*y[14]
 r14 = k14*y[1]*y[6]
 r15 = k15*y[3]
 r16 = k16*y[4]
 r17 = k17*y[4]
 r18 = k18*y[16]
 r19 = k19*y[16]
 r20 = k20*y[17]*y[6]
 r21 = k21*y[19]
 r22 = k22*y[19]
 r23 = k23*y[1]*y[4]
 r24 = k24*y[19]*y[1]
 r25 = k25*y[20]

 dy1  = -r1-r10-r14-r23-r24+
          r2+r3+r9+r11+r12+r22+r25
 dy2  = -r2-r3-r9-r12+r1+r21
 dy3  = -r15+r1+r17+r19+r22
 dy4  = -r2-r16-r17-r23+r15
 dy5  = -r3+r4+r4+r6+r7+r13+r20
 dy6  = -r6-r8-r14-r20+r3+r18+r18
 dy7  = -r4-r5-r6+r13
 dy8  = r4+r5+r6+r7
 dy9  = -r7-r8
 dy10 = -r12+r7+r9
 dy11 = -r9-r10+r8+r11
 dy12 = r9
 dy13 = -r11+r10
 dy14 = -r13+r12
 dy15 = r14
 dy16 = -r18-r19+r16
 dy17 = -r20
 dy18 = r20
 dy19 = -r21-r22-r24+r23+r25
 dy20 = -r25+r24
 SA[dy1,dy2,dy3,dy4,dy5,dy6,dy7,dy8,dy9,dy10,dy11,dy12,dy13,dy14,dy15,dy16,dy17,dy18,dy19,dy20]
end

u0 = zeros(20)
u0[2]  = 0.2
u0[4]  = 0.04
u0[7]  = 0.1
u0[8]  = 0.3
u0[9]  = 0.01
u0[17] = 0.007
u0 = SA[u0...]
prob = ODEProblem(f,u0,(0.0,60.0))
@btime sol = solve(prob,Rosenbrock23(chunk_size = Val{8}()),save_everystep=false)
@btime sol = solve(prob,Rodas5(chunk_size = Val{8}()),save_everystep=false)

# calc_W returns lu(W):
3.936 ms (672 allocations: 405.56 KiB)
3.092 ms (1678 allocations: 566.19 KiB)

# calc_W returns W:
12.170 ms (482 allocations: 814.94 KiB)
23.892 ms (922 allocations: 1.54 MiB)

# calc_W returns inv(W):
4.082 ms (216 allocations: 664.12 KiB)
2.995 ms (166 allocations: 496.88 KiB)

# Now
3.823 ms (672 allocations: 405.56 KiB)
2.813 ms (1678 allocations: 566.19 KiB)
```

**Comment by ChrisRackauckas** [comment]
Looks done. @YingboMa bad idea?

**Review by YingboMa — COMMENTED** [review]
The change looks good to me, but it seems error prone. We should refactor this.

**Comment by ChrisRackauckas** [comment]
Yeah it's kind of the hackiest hack of the west, but I wanted to finally analyze it all. I had a lot more tests with Newton methods, and it seems for those it just almost never made sense: anything where this was better Rosenbrock destroyed it anyways. So the main takeaway is that for sufficiently small stiff ODEs, single inv in Rosenbrocks (of higher order) is the best thing you can do. So it's hacking a bit to force that method to exist. But hey, it's the absolute best method for this domain that I could find, and it's a very crucial domain for the pharma work, so it deserves some performance hacking.

**Comment by YingboMa** [comment]
Maybe something like
```julia
struct StaticWOperator{isinv, T}
    W::T
    function StaticWOperator(W::T) where T
        isinv = size(W, 1) <= CUTOFF
        StaticWOperator{isinv, T}(isinv ? inv(W) : W)
    end
end
isinv(W::StaticWOperator{S}) where S = S
Base.:\(W::StaticWOperator, v) = isinv(W) ? W.W * v : W.W \ v
```

File: src/perform_step/rosenbrock_perform_step.jl [hunk 1]
```diff
@@ -260,6 +260,7 @@ end
   dT = calc_tderivative(integrator, cache)
 
   W = calc_W(integrator, cache, γ, repeat_step)
+
```
Comment (ChrisRackauckas) [inline src/perform_step/rosenbrock_perform_step.jl:263]:
```suggestion
```

**Review by YingboMa — APPROVED** [review]
LGTM

**Pull request merged**